src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -515,6 +499,28 @@ async def build_pex(
     )
 
 
+def _validate_metadata(metadata: LockfileMetadata, request: PexRequest, python_setup: PythonSetup):
+    if metadata.is_valid_for(
+        request.requirements.lockfile_hex_digest,
+        request.interpreter_constraints,
+        python_setup.interpreter_universe,
+    ):
+        return None
+
+    message_1 = f"Invalid lockfile for PEX request `{request.output_filename}`. "
+    message_2 = (
+        "If your requirements or interpreter constraints have changed, follow the "
+        "instructions in the header of the lockfile to regenerate it. Otherwise, ensure your interpreter "
+        "constraints are compatible with the constraints specified in the lockfile."
+    )
```

Comment:
How about this?

> The lockfile `{lockfile_path}` is stale because either the input requirements have changed or the interpreter constraints are not compatible. Please regenerate the lockfile by following the instructions in the header of the lockfile.

We can assume either case has happened, rather than making the user guess if it did.

--

This also suggests to me that we will soon want to change `is_valid_for` to have an enum return type, or split it out into several functions, so that we can make this error message more informative. Rather than saying "either x or y", we can definitively say 

> The requirements hash in {path/to/lockfile.txt} is not expected. Please either change back the input requirements or regenerate the lockfile by following the instructions in the header of the lockfile.
>
> Alternatively, if you are manually maintaining this lockfile, update the requirements hash to be 'abc132'.

Or a specific error about ICs.

But that's fine as a follow up, given how much we're going to be iterating on this code for user lockfiles (e.g. `platforms` metadata).

Reply:
I tried incorporating `{lockfile_path}` but it's not really reliable for the tool lockfile case (the file is called `lockfile.txt`, the lockfile path is `black_default_lockfile.txt` or similar).

The interpreter constraints needs different advice to requirements: if a given user's interpreter constraints setting is lagging what the project thinks the interpreter constraints should be, and the project generates a new lockfile, then the user should update their interpreter constraints and not regenerate the lockfile.

I'm thinking of a case -- like for us with M1 macbooks -- where we override our interpreter constraints. There's a better chance of that becoming stale while the people maintaining the lockfiles move forward. The correct fix here would definitely not be us updating our lockfiles.

Reply:
Re the enum return type for `is_valid_for`, I'm definitely in favour of returning something to help inform our validation code. We probably want something more than an enum, so that we can build up a full error message rather than failing for the first mismatch we have (let me think about how to do that)